src/main/java/bio/terra/workspace/service/resource/controlled/flight/CreateGcsBucketStep.java

```diff
@@ -1,28 +1,136 @@
+      ControlledGcsBucketResource resource,
+      AuthenticatedUserRequest userRequest) {
     this.crlService = crlService;
+    this.resource = resource;
+    this.userRequest = userRequest;
   }
 
   @Override
   public StepResult doStep(FlightContext flightContext)
       throws InterruptedException, RetryException {
+
+    final BucketInfo bucketInfo =
+        BucketInfo.newBuilder(resource.getBucketName())
+            .setLocation(resource.getLocation())
+            .setStorageClass(ApiConversions.toGcsApi(resource.getDefaultStorageClass()))
+            .setLifecycleRules(ApiConversions.toGcsApi(resource.getLifecycle()))
+            .build();
+
+    final StorageCow storageCow = crlService.createStorageCow(userRequest);
+    final BucketCow bucketCow = storageCow.create(bucketInfo);
+
     return StepResult.getStepResultSuccess();
   }
 
   @Override
   public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    final StorageCow storageCow = crlService.createStorageCow(userRequest);
+    final boolean deleted = storageCow.delete(resource.getBucketName());
+
     return StepResult.getStepResultSuccess();
   }
+
+  private static class ApiConversions {
+
+    private ApiConversions() {}
+
+    private static StorageClass toGcsApi(GoogleBucketDefaultStorageClass storageClass) {
+      switch (storageClass) {
+        case STANDARD:
+          return StorageClass.STANDARD;
+        case NEARLINE:
+          return StorageClass.NEARLINE;
+        case COLDLINE:
+          return StorageClass.COLDLINE;
+        case ARCHIVE:
+          return StorageClass.ARCHIVE;
+        default:
+          throw new IllegalStateException("Unrecognized storage class " + storageClass);
+      }
+    }
+
+    private static List<LifecycleRule> toGcsApi(GoogleBucketLifecycle lifecycle) {
+      return lifecycle.getRules().stream()
+          .map(ApiConversions::toGcsApi)
+          .collect(Collectors.toList());
+    }
+
+    private static LifecycleRule toGcsApi(GoogleBucketLifecycleRule lifecycleRule) {
+      return new LifecycleRule(
+          toGcsApi(lifecycleRule.getAction()), toGcsApi(lifecycleRule.getCondition()));
+    }
+
+    private static LifecycleAction toGcsApi(GoogleBucketLifecycleRuleAction lifecycleRuleAction) {
+      switch (lifecycleRuleAction.getType()) {
+        case DELETE:
+          return LifecycleAction.newDeleteAction();
+        case SET_STORAGE_CLASS:
+          return LifecycleAction.newSetStorageClassAction(
+              toGcsApi(lifecycleRuleAction.getStorageClass()));
+        default:
+          throw new IllegalStateException(
+              "Unrecognized lifecycle action type " + lifecycleRuleAction.getType());
+      }
+    }
+
+    private static LifecycleCondition toGcsApi(GoogleBucketLifecycleRuleCondition condition) {
+      final LifecycleCondition.Builder resultBuilder = LifecycleCondition.newBuilder();
+
+      // TODO: some conditions aren't on the Google api object
```

Comment:
Not sure what that means. Is there more work to do in this PR? If you are suggesting to do it later (which is fine), then maybe make a Jira ticket and put the ticket number on the TODO. I want to keep track of the loose ends.

Reply:
The list documented contained presumably obsolete conditions that don't seem to be available in the Google API for new buckets.

I'll check into this this afternoon.

Reply:
Turns out we need a new version of the storage library, meaning a new CRL release. It turns out there may be breakage in the new version, if the tests are to be believed, so I'm pulling this work into its own ticket so it doesn't block this thread.

Reply:
Work to update CRL and fill in the rest of these is [PF-506](https://broadworkbench.atlassian.net/browse/PF-506).